bestie/server/bigquery_metrics.go

```diff
@@ -0,0 +1,288 @@
+	project   string
+	dataset   string
+	tablename string
+}
+
+type keyValuePair struct {
+	key   string
+	value string
+}
+
+type testMetric struct {
+	metricname string
+	tags       []keyValuePair
+	value      float64
+	timestamp  int64
+}
+
+type metricTestResult struct {
+	table   bigQueryTable
+	metrics []testMetric
+}
+
+// BigQuery service location.
+const databaseLocation = "us-west1"
+
+// Define default BigQuery table to use if not specified in Bazel TestResult event message.
+//
+// NOTE: If BES Endpoint needs to shard metrics into multiple tables, it will add a suitable
+// suffix to the tablename specified here.
+var bigQueryTableDefault = bigQueryTable{
+	project:   "bestie-builds",
+	dataset:   "",            // Must be specified as --dataset arg on the command line.
+	tablename: "testmetrics", // Can be overridden from the --tablename arg on the command line.
+}
+
+// Save implements the ValueSaver interface.
+// This example disables best-effort de-duplication, which allows for higher throughput.
+func (i *bigQueryMetric) Save() (map[string]bq.Value, string, error) {
+	ret := map[string]bq.Value{
+		"metricname": i.metricname,
+		"tags":       i.tags,
+		"value":      i.value,
+		"timestamp":  i.timestamp,
+	}
+	return ret, bq.NoDedupeID, nil
+}
+
+func getMetricsTableSchema() bq.Schema {
+	databaseSchema := bq.Schema{
+		{Name: "metricname", Type: bq.StringFieldType, Description: "metric name", Required: true},
+		{Name: "tags", Type: bq.StringFieldType, Description: "metric attribute tags"},
+		{Name: "value", Type: bq.FloatFieldType, Description: "metric value"},
+		{Name: "timestamp", Type: bq.TimestampFieldType, Description: "sample collection timestamp", Required: true},
+	}
+	return databaseSchema
+}
+
+// Produce a formatted string of the table identifier.
+func (t bigQueryTable) formatDatasetId() string {
+	return fmt.Sprintf("`%s.%s`", t.project, t.dataset)
+}
+
+// Produce a formatted string of the dataset identifier.
+func (t bigQueryTable) formatTableId() string {
+	return fmt.Sprintf("`%s.%s.%s`", t.project, t.dataset, t.tablename)
+}
+
+// Check if a BigQuery table exists based on reading its metadata.
+func (t bigQueryTable) isTableExist(session bigQuerySession) bool {
+	_, err := session.client.Dataset(t.dataset).Table(t.tablename).Metadata(session.ctx)
+	return err == nil
+}
+
+// Normalize BigQuery table references by filling in with defaults, as needed.
+func normalizeTableRef(t *bigQueryTable) {
+	// Always use the default project.
+	t.project = bigQueryTableDefault.project
+
+	// If either of the following were not found in the protobuf message,
+	// their zero value will show up here.
+	if len(t.dataset) == 0 {
+		t.dataset = bigQueryTableDefault.dataset
+	}
+	if len(t.tablename) == 0 {
+		t.tablename = bigQueryTableDefault.tablename
+	}
+}
+
+// Create a new dataset using an explicit destination location.
+func createDataset(w io.Writer, t bigQueryTable, location string, session bigQuerySession) error {
```

Comment:
Usually something like this, e.g. S3 bucket should always exist, should be a Must type function.

Instead of returning the error, how about renaming ti something like MustCreateDataset() and if the error returned is not "already exists", panic.

Because the program can't run unless the bucket's been made, a panic is good.  Also it should be called as soon as possible in the program.

Reply:
I agree with the sentiment (have bestie require the dataset/tables to already exist, and error if they don't) - disagree with the implementation - don't panic in production code; instead, return the error to the top-level and have it exit gracefully. Since the function doesn't panic, don't prefix the name with `Must`.

We could either:
* check that the tables exist at startup, and error early if they don't
* error on every request if the table doesn't exist

...this depends partially on what's more convenient from the bigquery library, with the former being a little easier to deal with operationally.

Reply:
This negates the ability for a test client application to define its own dataset and/or tablename in the *.metrics.pb file as an indication of where it wants to store its metrics in BigQuery. My main concern is that I don't know how other parts of the org plan on storing their metrics, so I wanted to allow for an optional means of specifying their destination. Am I being too flexible with this?

Reply:
Update: Discussed this with Mike G. We like having the current options to specify a different dataset and/or tablename, but agree that it's reasonable to expect those to be pre-created manually to simplify the BES Endpoint in that regard. I'll probably do something like track non-existing "project.dataset.tablename" requests in a map, so that we can log the first occurrence and just bump a service counter on subsequent occurrences.